### PR TITLE
update tfx version to 0.21.4

### PR DESCRIPTION
**Pull request by hongye-sun**

### CODE REVIEW

**Comment by numerology** [comment]
@hongye-sun I think I see the problem you mentioned on why the requirements.txt won't be updated.

It seems like there is a recent dependency conflict introduced in TFX.
```
There are incompatible versions in the resolved dependencies:
  kubernetes==9.0.0 (from -r requirements.in (line 3))
  kubernetes<11,>=10.0.1 (from tfx==0.21.4->-r requirements.in (line 12))
```
pip-compile fails here. I think it's okay to commit as long as the tests passes.

**Comment by hongye-sun** [comment]
@numerology will there by any issue if the requirements.txt is not in sync?

**Comment by numerology** [comment]
> @numerology will there by any issue if the requirements.txt is not in sync?

I don't think so. As long as this current *frozen* set of dependencies can work, we're in good shape.

File: backend/requirements.in [hunk 1]
```diff
@@ -1 +1 @@
-tfx==0.21.2
+tfx==0.21.4
```
Comment (Ark-kun) [inline backend/requirements.in:1]: You'll need to regenerate requirements.txt too

File: test/sample-test/requirements.in [hunk 2]
```diff
@@ -9,4 +9,4 @@ google-api-python-client==1.7.8
 google-cloud-storage==1.17.0
 fire==0.2.1
 yamale==2.0
-tfx==0.21.2
+tfx==0.21.4
```
Comment (Ark-kun) [inline test/sample-test/requirements.in:12]: You'll need to regenerate requirements.txt too

Comment (numerology) [reply test/sample-test/requirements.in:12]: I tried myself, it seems like pip-compile is failing b/c the recent kubernetes incompatible dependency problem zhitao cc'ed you.

Comment (Ark-kun) [reply test/sample-test/requirements.in:12]: I think I've already fixed the Kubernetes problem in TFX.

This problem is different - for some reason this same `requirements.in` pins kubernetes package to a really old version. As far as I know, `requirements.in` should not pin versions like that - only `requirements.txt` should.

Comment (numerology) [reply test/sample-test/requirements.in:12]: I see. Thanks.

Comment (hongye-sun) [reply test/sample-test/requirements.in:6]: @Ark-kun @numerology is there any action item for this comment? I ran the update-requirements.sh but there is nothing is changed.

Comment (Ark-kun) [reply test/sample-test/requirements.in:6]: Please update/rebase to the latest master and run the improved script:
```
cd test/sample-test
cat ../../sdk/python/requirements.in requirements.in | ../../backend/update_requirements.sh google/cloud-sdk:279.0.0 >requirements.txt
```

Comment (Ark-kun) [reply test/sample-test/requirements.in:6]: I've also updated the `requirements.in` to remove some unneeded packages and relax the dependencies (e.g. `kubernetes`). See https://github.com/kubeflow/pipelines/pull/3536

File: backend/requirements.in (hunk 1 above)
Comment (Ark-kun) [reply backend/requirements.in:1]: Just to clarify: Changing only the `requirements.in` will not affect any build behavior. You'll need to update `requirements.txt` by running `./update_requirements.sh <requirements.in >requirements.txt`

File: test/sample-test/requirements.in (hunk 2 above)
Comment (hongye-sun) [reply test/sample-test/requirements.in:6]: @Ark-kun the command result in requirements.txt to be empty.

Comment (Ark-kun) [reply test/sample-test/requirements.in:6]: > @Ark-kun the command result in requirements.txt to be empty.

Looks like the requirements of `tfx==0.21.4` are self-inconsistent. I've fixed this issues in the TFX repo, but it's not release dyet.

Comment (hongye-sun) [reply test/sample-test/requirements.in:6]: I temporarily update the tfx version in requirements.txt file. If it passes E2E tests, we might go with the workaround to unblock the release.

**Comment by Ark-kun** [comment]
TFX 0.22.0 still has self-inconsistent dependencies:
```
Could not find a version that matches tensorflow!=2.0.*,!=2.2.0,<2.3,<3,>=1.15,~=2.2.0 (from tfx==0.22.0->-r -)
Tried: 0.12.0, 0.12.1, 1.0.0, 1.0.1, 1.1.0, 1.2.0, 1.2.1, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.7.0, 1.7.1, 1.8.0, 1.9.0, 1.10.0, 1.10.1, 1.11.0, 1.12.0, 1.12.2, 1.12.3, 1.13.1, 1.13.2, 1.14.0, 1.15.0, 1.15.2, 1.15.3, 2.0.0, 2.0.1, 2.0.2, 2.1.0, 2.1.1, 2.2.0
Skipped pre-versions: 0.12.0rc0, 0.12.0rc1, 1.1.0rc0, 1.1.0rc1, 1.1.0rc2, 1.2.0rc0, 1.2.0rc1, 1.2.0rc2, 1.3.0rc0, 1.3.0rc1, 1.3.0rc2, 1.4.0rc0, 1.4.0rc1, 1.5.0rc0, 1.5.0rc1, 1.6.0rc0, 1.6.0rc1, 1.7.0rc0, 1.7.0rc1, 1.8.0rc0, 1.8.0rc1, 1.9.0rc0, 1.9.0rc1, 1.9.0rc2, 1.10.0rc0, 1.10.0rc1, 1.11.0rc0, 1.11.0rc1, 1.11.0rc2, 1.12.0rc0, 1.12.0rc1, 1.12.0rc2, 1.13.0rc0, 1.13.0rc1, 1.13.0rc2, 1.14.0rc0, 1.14.0rc1, 1.15.0rc0, 1.15.0rc1, 1.15.0rc2, 1.15.0rc3, 2.0.0a0, 2.0.0b0, 2.0.0b1, 2.0.0rc0, 2.0.0rc1, 2.0.0rc2, 2.1.0rc0, 2.1.0rc1, 2.1.0rc2, 2.2.0rc0, 2.2.0rc1, 2.2.0rc2, 2.2.0rc3, 2.2.0rc4
There are incompatible versions in the resolved dependencies:
  tensorflow!=2.0.*,!=2.2.0,<3,>=1.15 (from ml-metadata==0.22.0->tfx==0.22.0->-r -)
  tensorflow!=2.0.*,<2.3,>=1.15 (from tensorflow-transform==0.22.0->tfx==0.22.0->-r -)
  tensorflow!=2.0.*,<3,>=1.15 (from tensorflow-data-validation==0.22.0->tfx==0.22.0->-r -)
  tensorflow!=2.0.*,<3,>=1.15 (from tensorflow-model-analysis==0.22.1->tfx==0.22.0->-r -)
  tensorflow!=2.0.*,<3,>=1.15 (from tfx==0.22.0->-r -)
  tensorflow<3,>=1.15 (from tfx-bsl==0.22.0->tfx==0.22.0->-r -)
  tensorflow~=2.2.0 (from tensorflow-serving-api==2.2.0->tfx==0.22.0->-r -)
```

`ml-metadata==0.22.0` is incompatible with `tensorflow-serving-api==2.2.0` regarding the tensorflow version.

**Comment by Ark-kun** [comment]
TFX has bee updated to 0.22.0

**Pull request closed without merging**